docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,491 @@
+management systems.
+
+## Motivation
+
+We need agreement on the API design direction and the semantics of the API to be able to
+implement and grow the flightctl service.
+
+### Goals
+
+* Define the format and behavior of the Device and Fleet APIs
+
+* Allow the management of individual fleetless devices on a smaller scale,
+  as an integration mechanism for services like AAP and ACM.
+
+* Allow the management of fleets of devices on a larger scale, via gitops or
+  API.
+
+* Change API fields to provide a more consistent experience.
+
+* Enable temporary/manual drift of fleet-managed devices for debug purposes.
+
+### Non-Goals
+
+
+
+## Proposal
+
+### Device API
+
+The device API is divided into multiple methods, some user facing, and other device/facing,
+this design facilitates cleaner RBAC control of the endpoint access. The diagram is only
+intended to illustrate the flow of information between the user/system and the device.
+
+```mermaid
+block-beta
+columns 1
+  user(["User"])
+
+  space
+
+
+  block
+    frontEndpoint["/device/{name}"]
+  end
+
+  user --> frontEndpoint
+  frontEndpoint --> user
+
+  space
+  block
+    metadata
+    devSpec["spec"]
+    devStatus["status"]
+  end
+
+  space
+
+  frontEndpoint --> metadata
+  frontEndpoint --> devSpec
+
+  block
+     renderedEndpoint["/device/{name}/rendered"]
+     statusEndpoint["/device/{name}/status"]
+  end
+
+  space
+
+  statusEndpoint --> devStatus
+  statusEndpoint --> devStatus
+  devSpec --> renderedEndpoint
+  renderedEndpoint --> Device
+  Device --> statusEndpoint
+
+
+  Device(["Device"])
+
+  style frontEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style renderedEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style statusEndpoint fill:#969,stroke:#333,stroke-width:4px
+  ```
+
+#### `/device/{name}` User facing method
+
+This endpoint can be accessed by the user or external systems to
+manage individual devices, also internally from the fleet specifications
+for devices.
+
+The proposed changes include:
+
+* Reincorporating device.spec as a visible part through this frontend API.
+
+* The `device.spec.os.image` contains a reference to the image to be used for the device.
+
+* The `device.spec.templateVersion` field is moved to `device.metadata.resourceVersion` opaque string identifies
+  the version of the device spec, this is used for tracking changes. This is moved from the metadata level, so
+  when we write in the spec field, the contents are then not modified by the server.
+
+* The `device.spec.{containers, systemd}` fields are moved under `device.spec.monitoring.{containers, systemd}`
+
+* The `device.spec.monitoring.containers` contains a list of container patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.monitoring.systemd` contains a list of systemd service patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.config` contains a list of config items
```

Comment:
list of zero or more?

Reply:
I assume that a list has that property: [] or [x] or [x,y,...], but adding it..

Reply:
Done.